doc/source/devel/array.rst

```diff
@@ -0,0 +1,612 @@
+header file.
+
+  * Defining the ``CHECK_BOUNDS`` macro, will cause checks of the validity of
+    indices every time an element is accessed and will raise an error when it
+    detects that an element that lies outside of the array bounds.
+
+  * Defining the ``CHECK_FINITE_ELEMENTS`` macro will cause a check during
+    retrieval of array elements that they are not ``NaN`` or ``inf``
+
+.. _array-examples:
+
+========
+Examples
+========
+
+Below, we show some factored out, simplified examples, ways in which how
+``CelloArray`` might simplify code:
+
+Copying Elements
+----------------
+
+This example illustrates how ``CelloArray`` simplifies the code
+required to copy elements between arrays. (We illustrate how one might
+write Nearest Neighbor reconstruction along the x-direction).
+
+This code assumes a mesh with shape ``(mz, my, mx)``. Suppose we have:
+
+  * An ``(mz,my,mx)`` array of cell-centered primitives ``w``
+
+  * An ``(mz,my,mx-1)`` array of left reconstructed values, ``wl``
+
+  * An ``(mz,my,mx-1)`` array of right reconstructed values, ``wr`` 
+
+First is an the ``CelloArray`` version:
+    
+.. code-block:: c++
+
+   typedef double enzo_float;
+   typedef CelloArray<enzo_float,3> EFlt3DArray;
+
+   void reconstruct_NN_x(EFlt3DArray &w, EFlt3DArray &wl, 
+                         EFlt3DArray &wr){
+       wl.subarray() = w.subarray(CSlice(0,w.shape(0)),
+                                  CSlice(0,w.shape(1)),
+                                  CSlice(0,-1));
+       wr.subarray() = w.subarray(CSlice(0,w.shape(0)),
+                                  CSlice(0,w.shape(1)),
+                                  CSlice(1,w.shape(2)));
+   }
+
+The analogous code using conventional pointer operations is:
+
+.. code-block:: c++
+
+   typedef double enzo_float;
+
+   void reconstruct_NN_x(enzo_float *w, enzo_float *wl, enzo_float *wr,
+                      int mx, int my, int mz){
+     int offset = 1;
+     for (int iz=0; iz<mz-1; iz++) {
+       for (int iy=0; iy<my-1; iy++) {
+         for(int ix=0; ix<mx-1; ix++) {
+           int i = (iz*my + iy)*mx + ix;
+           int i_xf = (iz*my + iy)*(mx-1) + ix; 
+           wl[i_xf] = w[i];
+           Wr[i_xf] = w[i + offset];
+         }
+       }
+     }
+   }
+
+Adding Flux Divergence
+----------------------
+
+We show a factored out, slightly simplified version of the code used
+to add the flux divergence in an unsplit manner. This example is one
+of the more notable cases where the ``CelloArray`` leads to more
+transparent code.
+
+This code assumes a mesh with shape (mz, my, mx). Suppose we have:
+
+  * An ``(mz,my,mx)`` array of cell-centered conserved quantities ``u``
+
+  * An ``(mz,my,mx-1)`` array of x-face centered fluxes in the x-direction,
+    ``xflux``
+
+  * An ``(mz,my-1,mx)`` array of y-face centered fluxes in the y-direction,
+    ``yflux``
+
+  * An ``(mz-1,my,mx)`` array of y-face centered fluxes in the z-direction,
+    ``zflux``
```

Comment:
I think that conceptually it'd be worth being a little more specific/verbose with respect to the dimensions here, e.g., saying that the `mx,my,mz` are the dimensions of the entire block including the ghost zones (which are not updated).
Otherwise, it is potentially confusing to use fluxes (at faces) with dimensions `-1` here and in the next example with the magnetic fields (also on the faces) using dimensions with `+1`